Adds ROCmProfilerService and fixes NVProfilerService.

ericcano opened the pull request:
#### PR description:

Splits the NVProfilerService into a generic framework activity annotation base and an NVTX specific customization. A ROCm customization is added. Another one for VTune should be easily derivable.

This commit also adds thread safety wit spinlocks to avoid thread scheduling effects. Some ranges experience double start of double end. This is now indicated with a mark, instead of failing on an assertion.

Support added for ES modules execution, path ranges, source and event cleanup (which is are the most important contributions to the inter event gap).

This could still be polished a lot, by adding configurable content or message, coloring by event/EDM stream, etc...

The support for EDM events is also not complete and could still be expanded.

#### PR validation:

The two services have been tested on NVIDIA and AMD GPUs and showed expected results:

Here with the PyTorch test:
<img width="1834" height="973" alt="image" src="https://github.com/user-attachments/assets/25f86886-2c53-4912-9d56-29c9f3a25f8a" />

<img width="1161" height="626" alt="image" src="https://github.com/user-attachments/assets/496ff457-a804-4409-8291-d8d6b1f7e1db" />

This was also tested with HLT configurations.

Commit:
Adds ROCmProfilerService and fixes NVProfilerService.

Splits the NVProfilerService into a generic framework activity annotation base and an NVTX specific
customization. A ROCm customization is added. Another one for VTune should be easily derivable.
This commit also creates TBB container based trackers for in flight
ranges, that match be begin/end of the range and wrap the undelying
range implementation. Coloring has been added (only supported in
NSight).

An option to annotate the ranges with may details is provided to beeter
accomodate a timeline view (full detail) and a statictical approach
(only the name of modules is added).

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"07b90deba69db3c0cb9b5a75f82bc865c50424b6":{"files":["PerfTools/ProfilerService/interface/ProfilerService.h","HeterogeneousCore/ROCmServices/plugins/BuildFile.xml","HeterogeneousCore/CUDAServices/plugins/BuildFile.xml"],"squashed":false,"time":1787743755},"0bcca00b6cb5e1166e69ed71bfb810c7c358f49b":{"files":["PerfTools/ProfilerService/interface/ProfilerService.h","HeterogeneousCore/ROCmServices/plugins/BuildFile.xml","HeterogeneousCore/CUDAServices/plugins/BuildFile.xml"],"squashed":false,"time":1787743755},"162c7bfcff4300227fa0392f13ac555a3798ce32":{"files":["PerfTools/ProfilerService/interface/ProfilerService.h","HeterogeneousCore/ROCmServices/plugins/BuildFile.xml","HeterogeneousCore/CUDAServices/plugins/BuildFile.xml"],"squashed":false,"time":1787743755},"2264189c78c304ebb606c15155d7e00eccd8ceb9":{"files":["HeterogeneousCore/CUDAServices/plugins/BuildFile.xml","HeterogeneousCore/CUDAServices/plugins/NVProfilerService.cc","HeterogeneousCore/ROCmServices/BuildFile.xml","HeterogeneousCore/ROCmServices/plugins/ROCmProfilerService.cc","PerfTools/ProfilerService/interface/ProfilerService.h","PerfTools/ProfilerService/interface/ProfilerServiceBase.h"],"squashed":true,"time":1783604801},"22c0e76f8b37baf68ed52a84ca5a632e9521d21c":{"files":["PerfTools/ProfilerService/interface/ProfilerService.h","HeterogeneousCore/ROCmServices/plugins/BuildFile.xml","HeterogeneousCore/CUDAServices/plugins/BuildFile.xml"],"squashed":false,"time":1787743754},"3038fa0c4e28ab31e0d93ee007355a96e9cdfcd6":{"files":["PerfTools/ProfilerService/interface/ProfilerService.h","HeterogeneousCore/ROCmServices/plugins/BuildFile.xml","HeterogeneousCore/CUDAServices/plugins/BuildFile.xml"],"squashed":false,"time":1787743755},"333e247cbb2a22f7fb6c73e02b2dca5efbe24915":{"files":["FWCore/Services/interface/ProfilerService.h","HeterogeneousCore/ROCmServices/plugins/BuildFile.xml"],"squashed":true,"time":1783599117},"3ee23b2f2e155d8a9a9e48bb31edaeb0f892447e":{"files":["HeterogeneousCore/ROCmServices/plugins/ROCmProfilerService.cc","PerfTools/ProfilerService/interface/ProfilerServiceBase.h"],"squashed":true,"time":1785996367},"418456fb8a1e85b22db880615683989f05b487e9":{"files":["PerfTools/ProfilerService/interface/ProfilerService.h","HeterogeneousCore/ROCmServices/plugins/BuildFile.xml","HeterogeneousCore/CUDAServices/plugins/BuildFile.xml"],"squashed":false,"time":1787743755},"56a27e6b4b1aa19e938ad7f035eb9d0997e69c4a":{"files":["HeterogeneousCore/CUDAServices/plugins/CUDAService.cc","HeterogeneousCore/CUDAServices/plugins/NVProfilerService.cc","HeterogeneousCore/ROCmServices/plugins/ROCmProfilerService.cc","PerfTools/ProfilerService/BuildFile.xml","PerfTools/ProfilerService/interface/ProfilerService.h","PerfTools/ProfilerService/interface/ProfilerServiceBase.h"],"squashed":true,"time":1783947245},"5c8bdbcf57ba53572610d7026764d0229f05dab7":{"files":["HeterogeneousCore/ROCmServices/BuildFile.xml"],"squashed":true,"time":1783948823},"5c8c78e5f6ab61d208bfa90f9eb33a9bd425b27d":{"files":["PerfTools/ProfilerService/BuildFile.xml"],"squashed":true,"time":1784124127},"6bdc6a19bd4196696e14b0e01f77d456636396fa":{"files":["PerfTools/ProfilerService/interface/ProfilerService.h","HeterogeneousCore/ROCmServices/plugins/BuildFile.xml","HeterogeneousCore/CUDAServices/plugins/BuildFile.xml"],"squashed":false,"time":1787743755},"78c590f5b73bf4c20ea06f70b180f866354a6d48":{"files":["HeterogeneousCore/CUDAServices/plugins/NVProfilerService.cc","HeterogeneousCore/CUDAServices/plugins/CUDAService.cc","HeterogeneousCore/ROCmServices/plugins/ROCmProfilerService.cc","FWCore/Services/interface/ProfilerServiceBase.h","FWCore/Services/interface/ProfilerService.h","HeterogeneousCore/CUDAServices/plugins/BuildFile.xml","HeterogeneousCore/ROCmServices/plugins/BuildFile.xml"],"squashed":true,"time":1783597196},"913189cb0b40c6a54f5dc531823785b604b12a77":{"files":["FWCore/Services/interface/ProfilerService.h","HeterogeneousCore/CUDAServices/plugins/BuildFile.xml","HeterogeneousCore/CUDAServices/plugins/NVProfilerService.cc","HeterogeneousCore/ROCmServices/plugins/BuildFile.xml","HeterogeneousCore/ROCmServices/plugins/ROCmProfilerService.cc"],"squashed":true,"time":1765282357},"a1a88c0f7afe78fdd5e0a4bed06d5eb6ee6f7045":{"files":["FWCore/Services/interface/ProfilerService.h"],"squashed":true,"time":1765283560},"a7794d396510170f577d2fec03453fa8db0bb651":{"files":["HeterogeneousCore/ROCmServices/plugins/ROCmProfilerService.cc"],"squashed":true,"time":1785998576},"b831b8c771c870fa654d60d24075d5daa595476b":{"files":["PerfTools/ProfilerService/interface/ProfilerService.h"],"squashed":true,"time":1785998359},"d1f90d6bd34cd97354c23ec8c69f159d46ca07b2":{"files":["HeterogeneousCore/ROCmServices/plugins/BuildFile.xml"],"squashed":true,"time":1765978740},"d7b622e8086c6496ea97ee7b814ef2ab79954b6f":{"files":["PerfTools/ProfilerService/interface/ProfilerService.h","HeterogeneousCore/ROCmServices/plugins/BuildFile.xml","HeterogeneousCore/CUDAServices/plugins/BuildFile.xml"],"squashed":false,"time":1787749045},"deed0e116996924cb2d6714fcc933259ff6992e3":{"files":["HeterogeneousCore/CUDAServices/plugins/BuildFile.xml","HeterogeneousCore/CUDAServices/plugins/NVProfilerService.cc","HeterogeneousCore/ROCmServices/BuildFile.xml","HeterogeneousCore/ROCmServices/plugins/BuildFile.xml","PerfTools/ProfilerService/interface/ProfilerServiceBase.h"],"squashed":false,"time":1787817979}},"emoji":{"4950767099":"+1","4959717474":"+1","5201404199":"+1","5201405333":"+1","5202185441":"+1","5202455052":"+1","5328069282":"+1"},"last_seen_sha":"deed0e116996924cb2d6714fcc933259ff6992e3","signatures":{"5202185441":"3ee23b2f2e155d8a9a9e48bb31edaeb0f892447e"}} -->

cmsbuild commented:
Pull request #49580 was updated.

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_20_0_X. Please open a backport if it should also go in to CMSSW_17_0_X.

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_20_1_X. Please open a backport if it should also go in to CMSSW_20_0_X.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49580/50141

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDAServices/plugins/BuildFile.xml  modified in PR(s): #51395
  - File  HeterogeneousCore/CUDAServices/plugins/CUDAService.cc  modified in PR(s): #51395

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49580/50141/code-checks.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49580/50141/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

ericcano commented:
The PR was heavily reworked. All the clashes (double starts and spurious end of ranges) are fixed. All signals are now covered. An option to have more info per range (better for timeline exploration) or less (for statistics) has been added. It is not ready so review.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49580/50142

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDAServices/plugins/BuildFile.xml  modified in PR(s): #51395
  - File  HeterogeneousCore/CUDAServices/plugins/CUDAService.cc  modified in PR(s): #51395

cmsbuild commented:
A new Pull Request was created by  @ericcano for master.

It involves the following packages:

- FWCore/Services (**core**)
- HeterogeneousCore/CUDAServices (**heterogeneous**)
- HeterogeneousCore/ROCmServices (**heterogeneous**)


@Dr15Jones, @cmsbuild, @fwyzard, @makortel, @smuzaffar can you please review it and eventually sign? Thanks.
@fwyzard, @makortel, @rovere, @wddgit this is something you requested to watch as well.
@ftenchini, @mandrenguyen, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

ericcano commented:
@felicepantaleo, you might be interested as this is similar to https://github.com/cms-sw/cmssw/pull/51271

Dr15Jones commented:
Given the files you put into `FWCore/Services` do not actually create a service (particular one that isn't needed by either the framework or by product) maybe a better place would be to put that code into the `PerfTools` subsystem?

Commit:
Moves shared ProfilerService headers to PerTools.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49580/50147

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDAServices/plugins/BuildFile.xml  modified in PR(s): #51395
  - File  HeterogeneousCore/CUDAServices/plugins/CUDAService.cc  modified in PR(s): #51395

cmsbuild commented:
Pull request #49580 was updated. @cmsbuild, @fwyzard, @makortel can you please check and sign again.

fwyzard commented on HeterogeneousCore/CUDAServices/plugins/CUDAService.cc:
```diff
@@ -7,6 +7,7 @@
 #include <string>
 #include <utility>
 #include <vector>
+#include <cstring>
```
This is already there on line 1.

fwyzard commented on HeterogeneousCore/CUDAServices/plugins/NVProfilerService.cc:
```diff
@@ -1,1235 +1,241 @@
 #include <vector>
+#include <utility>
+// For scoped locks
+#include <mutex>
```
Please remote the comment and sort alphabetically.

fwyzard commented on HeterogeneousCore/CUDAServices/plugins/NVProfilerService.cc:
```diff
@@ -1,1235 +1,241 @@
-      bool expected = false;
-      if (globalFirstEventDone_.compare_exchange_strong(expected, true))
-        cudaProfilerStart();
+ - the option '--cpu-profiling on' currently results in cmsRun being stuck at the beginning of the job.
+ - the option '--cpu-thread-tracing on' is not compatible with jemalloc, and should only be used with cmsRunGlibC.)";
```
please restore the original indentation

fwyzard commented:
assign core

I think we need a more comprehensive review of the approach to profiling, also in light of #51271 .

Independently, I'm wondering if having three or more services (CUDA, ROCm, eventually VTune, etc.) is the best approach, or if we should try to wrap their interface at a lower level.
https://github.com/alpaka-group/bactria aims to do something like that, but I think the architecture it uses wouldn't go along well with CMSSW.

cmsbuild commented:
New categories assigned: core

@Dr15Jones,@makortel,@smuzaffar you have been requested to review this Pull request/Issue and eventually sign? Thanks

Commit:
Inclusions cleanup and creation of BuildFile.xml

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49580/50178

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDAServices/plugins/BuildFile.xml  modified in PR(s): #51395

cmsbuild commented:
Pull request #49580 was updated. @Dr15Jones, @cmsbuild, @fwyzard, @makortel, @smuzaffar can you please check and sign again.

Commit:
Reinstates formerly present use and makes ProfilerService header only.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49580/50180

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDAServices/plugins/BuildFile.xml  modified in PR(s): #51395

cmsbuild commented:
Pull request #49580 was updated. @Dr15Jones, @cmsbuild, @fwyzard, @makortel, @smuzaffar can you please check and sign again.

ericcano commented:
> I think we need a more comprehensive review of the approach to profiling, also in light of #51271 .

I see there is resource tracking in the `PerfettoTraceService`, but the coverage of signals is less. We could probably join efforts, indeed. I could envision an additional member function(s) to the Range (to be renamed), which would support value tracking and have a noop for the backends that support it. NVTX seems to have them [in the pipeline](https://forums.developer.nvidia.com/t/nvtx-3-2-release-announcement/333012), but I don't think we have them yet in the current release.

Symetrically, the the `PerfettoTraceService` might inherit from the ProfilerService to get full signal coverage, and still add the counters on top.

> 
> Independently, I'm wondering if having three or more services (CUDA, ROCm, eventually VTune, etc.) is the best approach, or if we should try to wrap their interface at a lower level. https://github.com/alpaka-group/bactria aims to do something like that, but I think the architecture it uses wouldn't go along well with CMSSW.

The ideas here was to factorize out the repetitive instrumentation, but the environments of the 3 variants are quite different, and we could also imagine a timing tracker as well.  Having a single modules knowing all of them could be cumbersome.

PS: I fixed the inclusions (they could be pruned quite a lot), and added dependency information in a `BuildFile.xml`.

fwyzard commented:
please test

cmsbuild commented:
-1

**Failed Tests**:   ClangBuild 
**Size**: This PR adds an extra 36KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa145b/54729/summary.html
**COMMIT**: 5c8bdbcf57ba53572610d7026764d0229f05dab7
**CMSSW**: CMSSW_20_1_X_2026-07-13-1100/el9_amd64_gcc13
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/49580/54729/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Failed Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' /usr/bin/time -v scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

Commit:
Fixes extra entry in BuildFiles.xml

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49580/50209

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDAServices/plugins/BuildFile.xml  modified in PR(s): #51395

cmsbuild commented:
Pull request #49580 was updated. @Dr15Jones, @cmsbuild, @fwyzard, @makortel, @smuzaffar can you please check and sign again.

makortel commented:
> I think we need a more comprehensive review of the approach to profiling, also in light of #51271 .

Do you have any specific questions or concerns in mind?

The `ActivityRegistry` signals are general such that anyone can use them as they wish. 

The `ProfilerService` of this PR looks more of an abstraction of commonalities of `NVProfilerService` and `ROCmProfilerService` than a general-purpose utility. 

The purpose looks to me like it is really about annotating traces produced by external tracing tools. I think it would make the purpose of the common code clearer if that would be reflected in their names.

In connection to #51271 I can't really comment without doing a very deep dive, so in the leading order I'd leave it to @felicepantaleo and @ericcano to sort out if joint infrastructure makes sense or not.

> Independently, I'm wondering if having three or more services (CUDA, ROCm, eventually VTune, etc.) is the best approach, or if we should try to wrap their interface at a lower level. https://github.com/alpaka-group/bactria aims to do something like that, but I think the architecture it uses wouldn't go along well with CMSSW.

IIUC technology-specific plugins would be needed at some point anyhow, so the question looks like whether to use the framework component plugins (Services), or make one framework component that internally uses plugins. From the framework point of view using the framework's plugin mechanism directly would seem most natural. The system itself is generic enough though such that both approaches can be done.

fwyzard commented:
test parameters:
  - enable = gpu

fwyzard commented:
please test

cmsbuild commented:
-1

**Failed Tests**:   ClangBuild 
**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa145b/55123/summary.html
**COMMIT**: 5c8c78e5f6ab61d208bfa90f9eb33a9bd425b27d
**CMSSW**: CMSSW_20_1_X_2026-08-05-2300/el9_amd64_gcc13
**Additional Tests**: GPU,AMD_MI300X,AMD_W7900,NVIDIA_H100,NVIDIA_L40S,NVIDIA_T4
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/49580/55123/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Failed Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' /usr/bin/time -v scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

fwyzard commented:
@ericcano can you make clang happy ?

https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa145b/55123/clang-new-warnings.log

fwyzard commented:
@ericcano can you move
```c++
using namespace std::string_literals;
```
from the header file back to the .cc files ?

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49580/50473

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49580/50473/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49580/50473/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

fwyzard commented on PerfTools/ProfilerService/interface/ProfilerServiceBase.h:
```diff
@@ -0,0 +1,312 @@
+    void unlock() { flag_.clear(std::memory_order_release); }
+
+  private:
+    std::atomic_flag flag_;
```
No.

According to [cppreference](https://en.cppreference.com/cpp/atomic/atomic_flag/atomic_flag):
> ```c++
> constexpr atomic_flag() noexcept;
> ```
> Initializes `std::atomic_flag` to clear state.

fwyzard commented:
> Copilot uses AI. Check for mistakes.

Indeed.

Reviewing code is time-consuming enough that I don't need to spend time to read also correct-looking-but-possibly- **wrong** auto-generated pseudo-review comments.

fwyzard commented:
-heterogeneous

I refuse to review PRs with Copilot-generated comments.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49580/50474

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49580/50474/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49580/50474/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49580/50475

cmsbuild commented:
Pull request #49580 was updated. @Dr15Jones, @cmsbuild, @fwyzard, @makortel, @smuzaffar can you please check and sign again.

ericcano commented:
Apologies, I had turned this automatic AI review feature on some time ago. It's now off.
@fwyzard, indeed I had looked up exactly what it made you point out before committing. Sorry about that.

fwyzard commented:
please test

smuzaffar commented:
please test

cmsbuild commented:
-1

**Failed Tests**:   Build  ClangBuild 
**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa145b/55442/summary.html
**COMMIT**: a7794d396510170f577d2fec03453fa8db0bb651
**CMSSW**: CMSSW_20_1_X_2026-08-17-2300/el9_amd64_gcc14
**Additional Tests**: GPU,AMD_MI300X,AMD_W7900,NVIDIA_H100,NVIDIA_L40S,NVIDIA_T4
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/49580/55442/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Failed Build

I found compilation error when building: 

<pre>/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/gcc/14.3.1-724da22786638848892aa9ded8fcd995/bin/c++ -c -DCMS_MICRO_ARCH='x86-64-v3' -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=140301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DBOOST_MPL_IGNORE_PARENTHESES_WARNING -D__HIP_PLATFORM_HCC__ -D__HIP_PLATFORM_AMD__ -DCMSSW_GIT_HASH='CMSSW_20_1_X_2026-08-17-2300' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_20_1_X_2026-08-17-2300' -Isrc -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/cms/cmssw-patch/CMSSW_20_1_X_2026-08-17-2300/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/pcre/8.43-b1efa343d62cf66cc5cafcce5ddcf5b0/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocm-hip/7.14-7f2dec530de8715894eaa38d9d02337d/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/boost/1.91.0-69d2f6d010ebd6581aa6b7ae1c4d6dd7/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/bz2lib/1.0.8-59f9cf2f694ed1cbcfd89fedefa58cc3/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/libuuid/2.40-30397f5d4641a42f70e122f23e043a5a/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocr-runtime/7.14-2924950508623cb3499ea900c5f293e6/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/lcg/root/6.36.13-15104c8b817845d1891e93d3b5a4fbfe/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/tbb/v2022.3.0-fc86502c2a3e05fe0a93b7701254cc8e/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/xz/5.8.3-227f994f587b0d5d8ef0de48972ee444/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/zlib/1.3.2-611ad6ca171d38fc22fbd2e9579e0775/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/fmt/10.2.1-94b5da68daf94c2ef846640fd869fdc9/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/hiprand/7.14-2625b93445d63767683bb7b6f47c3073/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/md5/1.0.0-4c04fac6c42256d53fbdabb58aa703f3/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/numactl/2.0.19-cd4812a8566d36bb2870143cdf8cdb1a/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rccl/7.14-06bd9762a327720a29e7b268576cbfd7/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocm-comgr/7.14-ef8503d776b88c7af8ec79a54273b639/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocm-core/7.14-5e95e6c59661f0bce6ec4348e0f42ba4/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocm-rocprofiler-sdk/7.14-f54783ceb18779e33deb3e157c9c7759/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocm-smi-lib/7.14-be7a4a0bd360cac3f127cead2bc0f833/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocprim/7.14-569215d2bf917aea39a0791c4162305f/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocprofiler-register/7.14-eba767829a32897e4b6c9d8a0e022b69/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocrand/7.14-656099cbec33f54b2ec8c897a9887711/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocthrust/7.14-b0d0bc1eea1ea5d341d7ad97a0a2b5c9/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/tinyxml2/6.2.0-c4563d5fb92fc6565e3044fafb0c97d1/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=unused-value -Werror=unused-result -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el9_amd64_gcc14/src/HeterogeneousCore/ROCmServices/plugins/HeterogeneousCoreROCmServicesPlugins/ROCmProfilerService.cc.d src/HeterogeneousCore/ROCmServices/plugins/ROCmProfilerService.cc -o tmp/el9_amd64_gcc14/src/HeterogeneousCore/ROCmServices/plugins/HeterogeneousCoreROCmServicesPlugins/ROCmProfilerService.cc.o
>> Compiling edm plugin src/HeterogeneousCore/ROCmServices/plugins/ROCmService.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/gcc/14.3.1-724da22786638848892aa9ded8fcd995/bin/c++ -c -DCMS_MICRO_ARCH='x86-64-v3' -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=140301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DBOOST_MPL_IGNORE_PARENTHESES_WARNING -D__HIP_PLATFORM_HCC__ -D__HIP_PLATFORM_AMD__ -DCMSSW_GIT_HASH='CMSSW_20_1_X_2026-08-17-2300' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_20_1_X_2026-08-17-2300' -Isrc -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/cms/cmssw-patch/CMSSW_20_1_X_2026-08-17-2300/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/pcre/8.43-b1efa343d62cf66cc5cafcce5ddcf5b0/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocm-hip/7.14-7f2dec530de8715894eaa38d9d02337d/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/boost/1.91.0-69d2f6d010ebd6581aa6b7ae1c4d6dd7/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/bz2lib/1.0.8-59f9cf2f694ed1cbcfd89fedefa58cc3/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/libuuid/2.40-30397f5d4641a42f70e122f23e043a5a/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocr-runtime/7.14-2924950508623cb3499ea900c5f293e6/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/lcg/root/6.36.13-15104c8b817845d1891e93d3b5a4fbfe/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/tbb/v2022.3.0-fc86502c2a3e05fe0a93b7701254cc8e/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/xz/5.8.3-227f994f587b0d5d8ef0de48972ee444/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/zlib/1.3.2-611ad6ca171d38fc22fbd2e9579e0775/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/fmt/10.2.1-94b5da68daf94c2ef846640fd869fdc9/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/hiprand/7.14-2625b93445d63767683bb7b6f47c3073/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/md5/1.0.0-4c04fac6c42256d53fbdabb58aa703f3/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/numactl/2.0.19-cd4812a8566d36bb2870143cdf8cdb1a/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rccl/7.14-06bd9762a327720a29e7b268576cbfd7/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocm-comgr/7.14-ef8503d776b88c7af8ec79a54273b639/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocm-core/7.14-5e95e6c59661f0bce6ec4348e0f42ba4/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocm-rocprofiler-sdk/7.14-f54783ceb18779e33deb3e157c9c7759/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocm-smi-lib/7.14-be7a4a0bd360cac3f127cead2bc0f833/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocprim/7.14-569215d2bf917aea39a0791c4162305f/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocprofiler-register/7.14-eba767829a32897e4b6c9d8a0e022b69/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocrand/7.14-656099cbec33f54b2ec8c897a9887711/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/rocthrust/7.14-b0d0bc1eea1ea5d341d7ad97a0a2b5c9/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02955/el9_amd64_gcc14/external/tinyxml2/6.2.0-c4563d5fb92fc6565e3044fafb0c97d1/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=unused-value -Werror=unused-result -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el9_amd64_gcc14/src/HeterogeneousCore/ROCmServices/plugins/HeterogeneousCoreROCmServicesPlugins/ROCmService.cc.d src/HeterogeneousCore/ROCmServices/plugins/ROCmService.cc -o tmp/el9_amd64_gcc14/src/HeterogeneousCore/ROCmServices/plugins/HeterogeneousCoreROCmServicesPlugins/ROCmService.cc.o
In file included from src/HeterogeneousCore/ROCmServices/plugins/ROCmProfilerService.cc:5:
src/PerfTools/ProfilerService/interface/ProfilerService.h: In constructor 'ProfilerService<Backend>::ProfilerService(const edm::ParameterSet&, edm::ActivityRegistry&)':
src/PerfTools/ProfilerService/interface/ProfilerService.h:491:12: error: 'class edm::ActivityRegistry' has no member named 'watchJobFailure'
  491 |   registry.watchJobFailure(this, &ProfilerService::jobFailure);
      |            ^~~~~~~~~~~~~~~
src/PerfTools/ProfilerService/interface/ProfilerService.h:557:12: error: 'class edm::ActivityRegistry' has no member named 'watchPostESModuleRegistration'; did you mean 'watchPostESModuleConstruction'?
  557 |   registry.watchPostESModuleRegistration(this, &ProfilerService::postESModuleRegistration);
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
</pre>

## Failed Clang Build

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' /usr/bin/time -v scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Local Products Rules ..... done
>> Creating project symlinks
>> Entering Package HeterogeneousCore/CUDAServices
>> Entering Package HeterogeneousCore/ROCmServices
>> Compile sequence completed for CMSSW CMSSW_20_1_X_2026-08-17-2300
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 1
Command exited with non-zero status 1
	Command being timed: "scram build -k -j 32 COMPILER=llvm compile BUILD_LOG=yes"
	User time (seconds): 29.39
	System time (seconds): 5.57
	Percent of CPU this job got: 373%
</pre>

makortel commented on PerfTools/ProfilerService/interface/ProfilerService.h:
```diff
@@ -0,0 +1,1304 @@
+  REGISTER_SIGNAL_WATCHER(BeginStream)
+  REGISTER_SIGNAL_WATCHER(EndStream)
+
+  registry.watchJobFailure(this, &ProfilerService::jobFailure);
```
This signal was removed in https://github.com/cms-sw/cmssw/pull/51626 (it has never been issued by the framework, and only in a unit test up to April 2007).

makortel commented on PerfTools/ProfilerService/interface/ProfilerService.h:
```diff
@@ -0,0 +1,1304 @@
+  REGISTER_SIGNAL_WATCHER(ESModuleConstruction)
+
+  // ES signal watchers
+  registry.watchPostESModuleRegistration(this, &ProfilerService::postESModuleRegistration);
```
This signal was also removed in https://github.com/cms-sw/cmssw/pull/51626. The `PostESModuleConstruction` signal (which apparently is already being watched above) serves the same purpose as this signal did.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49580/50726

cmsbuild commented:
Pull request #49580 was updated. @Dr15Jones, @cmsbuild, @fwyzard, @makortel, @smuzaffar can you please check and sign again.

m-fila commented on PerfTools/ProfilerService/interface/ProfilerServiceBase.h:
```diff
@@ -0,0 +1,312 @@
+      }
+    }
+
+    void unlock_shared() { state_.fetch_sub(1, std::memory_order_release); }
```
the mechanism for lock shared allow for starving writers - if there is enough new readers the writer might never get a chance to acquire.

This is probably not an issue since hopefully its used only in low-contention scenarios.
But is there a reason to roll a custom implementation of spinlocks rather than take them from tbb?

ericcano replied:
Indeed, switching to the TBB implementation resolves all those issues in one go. Switching to it.

m-fila commented on PerfTools/ProfilerService/interface/ProfilerServiceBase.h:
```diff
@@ -0,0 +1,312 @@
+  };
+
+  /// @brief Reader-writer spinlock.
+  /// Compatible with std::lock_guard (exclusive) and std::shared_lock (shared).
```
`std::shared_lock` requires a type to fulfill SharedLockable named requirement which this class doesn't meet. It's a bit pedantic but all the usage of `shared_lock` with `RWSpinLock` is non-conforming code.

ericcano replied:
Fixed with TBB's.

m-fila commented on PerfTools/ProfilerService/interface/ProfilerServiceBase.h:
```diff
@@ -0,0 +1,312 @@
+  /**
+    * @brief Abstract color enumeration the derived classes can translate (or disregard).
+    */
+  class SpinLock {
```
Shouldn't this be a docstring for spinlock?

ericcano replied:
Fixed with TBB's.

m-fila commented on PerfTools/ProfilerService/interface/ProfilerServiceBase.h:
```diff
@@ -0,0 +1,312 @@
+  public:
+    RangePool() : next_allocation_size_(kInitialAllocationSize) { allocateUnlocked_(kInitialAllocationSize); }
+
+    size_t acquireSlot() {
```
```suggestion
    std::size_t acquireSlot() {
```
in several places

m-fila commented on PerfTools/ProfilerService/interface/ProfilerServiceBase.h:
```diff
@@ -0,0 +1,312 @@
+    RWSpinLock mutex_;
+    RangePool<Range>& range_pool_;
+    bool show_detailed_info_;
+    tbb::concurrent_unordered_map<Key, size_t, boost::hash<Key>> in_flight_;
```
`tbb::concurrent_hash_map` supports concurrent erasure, which would potentially simplify code (no need to create a separate storage for ranges and indices).
Was there a reason to use `tbb::concurrent_unordered_map` instead?

ericcano replied:
The hash map has its own internal locking to achieve erasure safety, and we need our own locking anyway for wider-than-the-map operations (like when allocating extra Ranges), so the manual locking avoids duplication.

ericcano replied:
... also the separate storage for Ranges and indices and the exponential pre-allocation minimizes memory allocations during signal handling, to make it as little intrusive as possible, and they are recycled via a non blocking FIFO.

m-fila replied:
Yeah, I get it. I was just wondering whether  you tried a simpler implementation first and found it not good enough or just went ahead and started optimizing it straight away

ericcano replied:
I don't recall a specific issue beyond limiting allocation.

m-fila replied:
Thanks

m-fila commented on HeterogeneousCore/CUDAServices/plugins/NVProfilerService.cc:
```diff
@@ -1,1235 +1,227 @@
-
 #include <cuda_profiler_api.h>
 #include <nvtx3/nvToolsExt.h>
+#include <nvtx3/nvtx3.hpp>
```
Is the header for NVTX C++ needed? Perhaps I missed but it seems that only C API is used.

ericcano replied:
Yes, it was extraneous.

m-fila reviewed (commented):
Hi Eric, I have a few comments and curiosities if you don't mind

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49580/50737

cmsbuild commented:
Pull request #49580 was updated. @Dr15Jones, @cmsbuild, @fwyzard, @makortel, @smuzaffar can you please check and sign again.
